### ISSUE TITLE
Use an object to store modifiers associated with a type of card / item effect

### ISSUE THREAD

**Issue by Turtyo** [Saplings-Projects/1M_sub#7]
## Feature Description

An object that holds the modifications to the state of the game that change the numerical values of effects. Everything modifies and checks this object when the need to make a calculation that can be impacted by buff / debuff arises.

## Motivation

This allows for a centralized modifier checking system when calculating damages. If I have an item that allows me to `heal +5` on every heal, every card that heals me shouldn't have to know that I have said object. Instead the item modifies this object that holds the changes, and every card then checks the same dictionary to know if there is a modification.

## Acceptance Criteria

- The object can be modified by buff / debuff
- The object can hold new types of effects easily
- The object holds information for both permanent changes (changes due to objects that stay after the end of a combat) and changes from temporary effects (that end after a set number of turns / end of combat)

## Proposed Solution

A dictionary of effects and the associated buffs / debuff that goes with them. For example, if we have `slash` damage and `piercing` damage:
```
{
    "slash" :  { Add: 10, Multiply: 2}
    "piercing" :  { Add: -15, Multiply: 1}
}
```
Would make a card that deals 10 `slash` damages and 20 `piercing` damages a total of `(10 + 10)*2 = 40 slash` damages  and `(20 - 15)*1 = 5 piercing` damages.

The problem with the previous implementation is that it doesn't differentiate between permanent and temporary status changes. This can be solved by having the following structure for the value associated to a given key:
`(Add permanent: value_1, Add temporary: value_2, Multiply permanent: value_3, Multiply temporary: value_4)`
Where the temporary values would be reset to 0 after each combat and the permanent values would only change if something adds or removes an item that changes this value.

## Additional Context

This overrules [Use a `Dictionary` to store buffs instead of an `Array`](https://github.com/Saplings-Projects/1M_sub/issues/5) by being a more general approach to the problem of storing buffs / debuffs. This is also in link with [Make stats changes related to character instead of calculating them at attack time](https://github.com/Saplings-Projects/1M_sub/issues/4)

**Comment by Mikotey** [comment]
Motivate:
> when calculating damages.
Just a nitpick. It's not exclusively for damage types. They could be used for anything including increasing summons spawned to increasing the # of items you get as a reward

Proposed Solution:
> A dictionary of effects and the associated buffs / debuff that goes with them
> This overrules [Use a Dictionary to store buffs instead of an Array](https://github.com/Saplings-Projects/1M_sub/issues/7#5) by being a more general approach to the problem of storing buffs / debuffs
The original intention wouldn't storing any numeric values like de/buffs. This is just consolidating the symptoms of buffs. So if the player has "Attack Up" and "Frenzy" (deal more, but take more), they could both communicate to the "Damage" key. I suppose we could allow it to store number values, but I feel like it could get messy with both on the same object

**Comment by Mikotey** [comment]
A couple of improvements I was thinking when discussing the system:

1. Have 2 modifiers objects. One for dealing effects and one for taking effects. So we wouldn't have to make a "Poison" modifiers and "PoisonResist" modifier. And the card can calculate the value by using the same "Poison" key from the dealer's modifiers and the receiver's modifiers
2. Create a property/resource class on the object that can associate keys as subcategories. This is mainly because of "Damage". A "slash" damage card should still see the generic damage increases of say "Attack Up" buff (a generic damage increasing buff). This could also be used to subcategorize ailments, like "thorns" and "entangle" are both "plant" debuffs

**Comment by Turtyo** [comment]
> The original intention wouldn't store any numeric values like de/buffs.
If we are using a dictionary, you would have a pair (key,value). If the value isn't a numerical one to signify a modification in the effects, what would it store then ?

1. I'm not exactly sure to understand. We would have two objects (let's say `DealingModifiers` and `ReceivingModifiers`, naming not great but we'll go with that for now). When dealing effects (like damage ?) you would look in the `DealingModifiers` and when receiving effects (again damage), you would look in `ReceivingModifiers`, and for example damage would have a `+2` modifier (if we have a card that says give `+2` defense for example ?). What would be the goal of this ?

2. So we would have a hierarchy of tags ? Why not directly use the parents of the buff class in itself ? Well I guess not every tag necessarily has an associated buff... But it could feel redundant with the buff hierarchy in itself

**Comment by Mikotey** [comment]
It is using numbers, but these numbers by themselves don't mean anything. It's like they are coefficients in math. Example "2X=?", we don't know the resulting value because we don't have X, but X is being doubled.

1. It's so we don't have to make extra variable names for receiving effects of an entity. Example, player plays a card with  "Apply 1 Poison to target". Player has a Rat Poison item giving modifier +3 to Poison applied. Card on calculation uses "Poison" as the key to check player's modifiers and sees the +3. However, the target has Poison Resistance with a modifier x0.5. When hovering target (or action resolution), it can use the same "Poison" key to check for receiving modifiers and see the x0.5. The result is only 2 stacks of Poison debuff is applied to the target.

2. The more I mentally talk circles explaining the flexibility that could be added, the more that I realize it is probably overkill. If someone wants to make a "Strength" buff that increases "slash", "blunt", and "pierce" damage, code it into the buff.  If they want to add a new physical damage type "slap", update the buff to include "slap". There'll likely be only 1 buff file that would be considering such collections

**Comment by Mikotey** [comment]
Okie, we can do that. So we'll add BuffCount to the custom class so it would look something like:
{
BuffCount : int = 0
AddPerm : int = 0 //do we want this as a float in the off-chance people want some kind of incremental bonus, like +1 every 5 kills gives 0.2 or something? just as an option
MultiplyPerm : float = 0
AddTemp : int = 0
MultiplyTemp : float = 0
}

1. essentially

**Comment by Mikotey** [comment]
That is holding the amount of a buff that is currently applied to you. This is mostly an issue with key reuse.

Example, poison. Theoretically, you can have poison, you can deal poison, and you can receive poison. We can differentiate the deal and receive with the "attack" vs "defense" data handling mentioned above. But how do you differentiate the debuff from the modifiers?
It is a bit strange since this value will exist on both the "attack" and "defense" dictionaries. Could probably redundantly set both through the object wrapper.
There's also the question if you would want a "temp" buff count similar to the temp modifiers.

**Comment by Turtyo** [comment]
What do you mean by "you can have poison" ? I understand dealing and receiving poison but that's pretty much the two states you can have on a character.
And I don't understand why you want to differentiate debuff vs modifiers, since debuff are status modifiers.

Also if I understood, even if `BuffCount` is both on the attack and defense dictionaries, those are separate object so the value won't be shared (unless we explicitly reference the number of effects as a value outside the dictionary and modify it outside, but I don't really see the point).

We could have a temp buff counter, but same as the buff counter, apart from having the data I don't really see a use to it.

**Comment by Turtyo** [comment]
So this dictionary would not only store modifiers but also status such as poison (which decreases once per turn then). If we do it like that, we would need something to check every buff in the array to see if they must decrease each turn. Which is why you might be suggesting the "temp" buff, but we would still have to check for those. Which would defeat the purpose of not having every key.
Or we can have a separate array / dict of status that have an effect every turn, so we don't need to check in the modifier dictionary for this.
When you mean both data structures, you mean both dict (the buff and debuff one ?). We could also keep the values separate and the total number of status would be the sum of the buff and debuff.

**Comment by Mikotey** [comment]
Ideally, no. I would prefer it to only house modifiers. Have the buff/debuffs be managed elsewhere, only to communicate with this object to communicate their modifications.
I've been interpreting this conversation that you want to house the buff/debuffs to be housed in the dictionary as well. Though, considering you're referring to poison as a "status", we must not have the same definition of what a buff/debuff is. When I refer to a "buff", I mean any persisting effect that provides a benefit. These don't have to be stat-related. They can include things like "glow" if in a game with darkness. Conversely, a "debuff" is any persisting effect that applies a negative effect. Also, doesn't have to be stat-related. Stun is a (generally short) debuff that usually doesn't affect stats, but skips the entities turn.
Rereading the conversation with that in mind. It sounds like we're both saying the same thing

I agree. We should have the de/buff/ailments/status effects held separately or at least in a separate data structure.

I mean it will house both the "attack/dealing" modifiers dictionary and the "defense/receiving" modifiers dictionary

**Comment by Turtyo** [comment]
The we agree with having buff elsewhere. For me the poison is a status, that's a debuff also, since it's a persisting effect (for a given number of turns) that applies a negative effect (damages).

So to sum it up:

Each entity in the scene has two dictionaries of the following shape:
```
{
    "slash" :  { BuffCountSlash: int = 5, PermanentAdd: int = 10, PermanentMultiply: float = 2, TempAdd: int = 1, TempMultiply: float = 0.2}
    "piercing" :  { BuffCountPiercing: int = 1, ...}
}
```
With one dictionary being for all the attack modifiers (everything that makes you do more or less damage), and the other for all the defense modifiers (everything that makes you take more or less damage).

I wouldn't have the `BuffCount` for each type be shared between the attack and defense dictionaries.

Separately, we will have another data structure (something like an Array), that holds the `Status` (for example `Poison`, which applies at the start of every turn, or `Weakness`, which reduces the damage of a target. Note that for weakness, the condition `Weakness` is in this Array, but the effect of the condition is being stored in the Attack dictionary that has been described above)

Do we agree on this @Mikotey ?

**Comment by Mikotey** [comment]
We don't need to have "buffCount" (aka status stacks) in the dictionary. We should have it in the separate data structure that holds the de/buff/statuses. Also don't have to worry about it being redundant in the two dictionaries.

But yes, I wholly agree with your "Weakness" example. We can iterate through this structure firing various events like "on turn start" and what not. Should we have that be a separate task?

**Comment by Turtyo** [comment]
Yes this should be a separate task in itself, but it will probably have to be done after #14 
We could name the structure for each entity as `entity_modifiers` for example. I was giving `buffCount` in the dictionary to have a count for each possible stat that can be modified.

**Comment by Turtyo** [comment]
Actually, the `Status` are already stored in an array and are launched at turn start or turn end. We will just need to modify the Status to change the dictionaries if needed.

**Comment by Mikotey** [comment]
Alrighty, sounds good 👍

**Comment by Turtyo** [comment]
I will try to finish the other issue I'm on for now. If no one assigned themselves to this issue in the meantime, then I'll do it

**Comment by Turtyo** [comment]
I will take this one since my other PR is pretty close to being merged i think (unless ?)

**Comment by Turtyo** [comment]
A little something I thought of:

To remove the correct amount of stats modifiers from a status when it ends, we need to store the amount by which it was changed originally (which we can do without problem with the "power" value that already exists in `StatusBase`)
Currently, if a status is already applied to an enemy, when applying it again we just add to the duration inside the array that stores all the currently applied Status

What if we have a status that adds 3 damages, and then we get another buff that buffs this status, thus making it boost our damage to 4 instead of 3. We apply the status, what about the amount ? does it change ? does it stay the same ? For example, if you have 10 turns of the boost 3, but you apply the modified status that gives you 4, does it add to the 3 thus becoming 7 for the number of turn left ? Does it take the number of turns left and use the new value of 4 instead of 3 (basically upgrading the already applied status) ?
does such a buff exist in the first place ?
can we buff a status, or do status always apply the same amount ?
do we only allow status modifications on permanent modifiers (eg no modifications of the status in themselves during the combat)

**Comment by Mikotey** [comment]
That's something for the buffs to manage. Set up some common events for buffs to trigger and handle themselves with local variables. OnApply(int), OnRemove(int), OnClear(), OnReapply(int), OnTurnStart(), OnTurnEnd(), OnAttack(), etc.

Buffs come in all kinds, some scale, some don't. An example is "Bleed" in some games is a fixed % damage, so adding more stacks just increases the duration. Many games, reapplication of a status will just refresh its duration. For value-based buffs like your boost example, upgrading/overwriting the current value can happen.

The flexibility is on the designer to do what they want with it. Just make sure the events are called properly and the designer sets the buff to clean itself up properly, or else you could get like a "memory leak" of buff modifiers.

> can we buff a status

You can set a buff to use a keyword like card effects. "Regeneration" could check the "Heal" key to see if there's any modifiers to the healing value before applying the heal

**Comment by Turtyo** [comment]
For all the OnApply, OnRemove etc... there is no problem, I see how to do it already, the question was on the change of value for a buff. i guess I will use a change in value and keep the number of turns. That's a choice in design...
